specification/draft-2/README.md

```diff
@@ -0,0 +1,687 @@
+TODO: Intro
+
+ - CWL specifies a way of describing runnable command line tools and workflows.
+ - CWL descriptions are serialized as either YAML or JSON.
+ - A CWL description contains a formal declaration of the inputs and outputs of described process (using Apache Avro schema).
+
+
+# Core specification
+
+CWL specification comes in two parts:
+
+ - Core spec allows description of basic tools and workflows
+ - Extensions add features that enable describing more complex tools and workflows
+
+If using an extension feature, it needs to be explicitly listed in the requirements section
+of the description document. Vendors might also add features incompatible with rest of CWL -
+in this case, they should be explicitly listed as requirements with full URI as value of `class` property.
+
+
+## Command line tools
+
+Basic tool example:
+
+```yaml
+class: CommandLineTool
+label: Grep
+description: Filter lines of text file by regular expression.
+inputs:
+  - id: pattern
+    type: string
+    binding:
+      position: 1
+  - id: in_file
+    type: File
+    streamable: true
+    binding:
+      position: 2
+outputs:
+  - id: out_file
+    type: File
+    streamable: true
+    binding:
+      glob: output.txt
+baseCmd: grep
+stdout: output.txt
+successCodes: [0, 1]
+```
+
+Above is defined a command line tool labeled "grep" with two required inputs
+(a file and a string) and one file output.
+
+Input objects contain a `binding` section which describes how the arguments
+are passed onto the command line. In this case, `pattern` and `in_file` are
+positional arguments and `pattern` is passed first, followed by `in_file`.
+Identifiers for inputs and outputs must be unique across the document.
+
+The `binding` section of the output object contains a single property `glob`
+with a glob pattern describing how to recognize file generated by the tool.
+
+Additionally, `baseCmd` and `stdout` properties describe the executable name
```

Comment:
Strongly disagree about running all commands through the shell, and in fact, the CommandLineTool is engineered specifically to avoid requiring that all commands run through the shell.  It doesn't mean you can't use the shell, you are just required to be explicit about it, e.g. ["/bin/sh", "-c", "first_cmd && second_cmd"] or use a wrapper script.

a) Putting it on users to quote things properly basically means that most scripts will fall over when they see a file with a space in it because getting quoting correct for all cases is very hard.  If the runtime is doing quoting for you, then it needs to have the command line already broken down into a list of tokens, at which point you might as well run the command directly.
b) It exposes potential security vulnerability to shell injection attacks, due to quoting mistakes and bugs such as [shellshock](https://en.wikipedia.org/wiki/Shellshock_%28software_bug%29).  A CWL tool that's a backend to a web service could be running on untrusted inputs.
c) Running the shell adds a little bit of (unnecessary) overhead compared to invoking the tool directly.